### PR TITLE
KAFKA-7909: Ensure timely rebalance completion after pending members rejoin or fail

### PR DESCRIPTION
**Pull request by wicknicks**
Fix the following situations, where pending members (one that has a member-id, and has requested to join a group) can cause rebalance operations to fail: 

- In AbstractCoordinator, a pending consumer should be allowed to leave.
- A rebalance operation must successfully complete if a pending member either joins or times out.
- During a rebalance operation, a pending member must be able to leave a group.

Signed-off-by: Arjun Satish <arjun@confluent.io>

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by ijuma** [comment]
Thanks for the PR. We should add a unit test.

**Comment by wicknicks** [comment]
retest this please.

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala
Comment (hachikuji) [inline core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala:789]: For logical consistency, I'd suggest this line go above `maybePrepareRebalance`. Removal of the pending member may resolve the rebalance.

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala
Comment (hachikuji) [inline core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala]: I would suggest something like this:
```scala
        if (group.is(PreparingRebalance))
          joinPurgatory.checkAndComplete(GroupKey(group.groupId))
```
We already have this logic inside `doJoinGroup` down one path. Maybe we can pull the check into a common path?

There are a couple other things we need to fix:

1. Inside `tryCompleteHeartbeat`, we need to call `forceComplete` if the pending member has joined.
2. Inside `onExpireHeartbeat`, we need to check for rebalance completion after removal of the pending member. Basically we need logic similar to `removeMemberAndUpdateGroup`.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: I think it is a bit confusing to call the generation valid if we are in the pending state. I am debating whether it is worth the complexity to add another `MemberState`, say `PENDING` for the state in which we have a memberId, but are not yet part of the group. Perhaps it would be sufficient to use something like this instead:
```java
public boolean hasMemberId() {
  return !memberId.isEmpty();
}
```

**Review by hachikuji — COMMENTED** [review]
Left a few comments.

**Comment by hachikuji** [comment]
cc @abbccdda

**Comment by abbccdda** [comment]
Thanks for the notice! @hachikuji

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala
Comment (abbccdda) [reply core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala]: +1 for using joinPurgatory.

**Comment by wicknicks** [comment]
@hachikuji @abbccdda thanks for the comments. this is ready for another pass.

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala
Comment (hachikuji) [inline core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala]: I am not sure this is right. It is the rebalance that may be blocked by this pending member. I think we want to do the same thing we did above.
```
if (group.is(PreparingRebalance)) {
  joinPurgatory.checkAndComplete(GroupKey(group.groupId))
}
```
By the way, are there any tests we could write for some of these fixes?

**Comment by wicknicks** [comment]
@hachikuji updated the PR (apologies for the confusion). will try to add some tests. thanks!

**Review by hachikuji — COMMENTED** [review]
The fixes LGTM. I'll await the test cases before merging. 

@abbccdda If you have time, would be nice to get a second review.

**Review by thatb96 — APPROVED** [review]

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:819]: How much effort would it be to have a test case for this? We have a few  LeaveGroup tests in `ConsumerCoordinatorTest`.

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala
Comment (hachikuji) [inline core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala]: I'm not sure how useful it is to log the fact that we're checking completion, but if you disagree, we should at least add the groupId to the message. Similarly the log line above on 921 should include the groupId.

File: core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala
Comment (hachikuji) [inline core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala:798]: A brief comment describing the test scenario would be helpful since these tests are complicated.

**Review by hachikuji — COMMENTED** [review]
Thanks for the updates. Left a few minor comments. We should also change the title of the PR. This will fix the connect test, but the main impact is fixing pending member completion and expiration.

File: core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala
Comment (wicknicks) [reply core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala:798]: comment added above.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java
Comment (wicknicks) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:819]: this has been added. resolving comment.

**Comment by wicknicks** [comment]
@hachikuji I've addressed your comments. Hopefully, this is good to be merged now. Thanks!

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java
Comment (hachikuji) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java]: A better test case would be to return the MEMBER_ID_REQUIRED error code. We could then skip SyncGroup and call maybeLeaveGroup. This is the scenario that this patch is trying to fix.

File: core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala
Comment (hachikuji) [inline core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala]: typo: "hearbeats"

File: core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala
Comment (hachikuji) [inline core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala]: nit: could be `val`

File: core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala
Comment (hachikuji) [inline core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala]: The group does not transition to `PreparingRebalance` until the pending member joins a second time with its assigned memberId. So there should be no reason for this JoinGroup if the pending member is fated to timeout as in this test case. I think we can just do the following after the `joinGroupPartial`.

1. Assert the group is stable.
2. Advance the clock to timeout the pending member.
3. Assert there are no pending members.
4. Assert the group is still stable.

File: core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala
Comment (hachikuji) [inline core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala]: Can we have a comment for this test case also?

File: core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala
Comment (hachikuji) [inline core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala]: nit: could be `val`

File: core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala
Comment (hachikuji) [inline core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala:962]: nit: it's conventional to handle `Some` first.

File: core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala
Comment (hachikuji) [inline core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala]: I don't really understand why we call `joinGroupPartial` twice. Wouldn't this create a fourth member?

File: core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala
Comment (hachikuji) [inline core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala]: Joining partially creates a pending member, but the rebalance won't be triggered until the pending member joins. To make the expectation clear, I'd suggest the following ordering:

1. Rejoin the group with the first member.
2. Assert the state is PreparingRebalance
3. Join with a pending member.
4. Assert numPending is 1
5. Join with the second member.
6. Assert the state is still PreparingRebalance
7. Timeout the pending member
8. Assert the state is CompletingRebalance

**Review by hachikuji — COMMENTED** [review]
Thanks for the test cases @wicknicks. I think we're almost there, just a few comments.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java
Comment (wicknicks) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java]: @hachikuji how can we mock this setup? `joinGroupIfNeeded` doesn't return until the JoinGroup has completed all the way to a successful SyncGroup. Is there a way to abort the future, or mark it as done in this test? I couldn't find an easy way to do this, that's why we went ahead with the adding the SyncGroup call. Also, returning MEMBER_ID_REQUIRED forces the second join group request to be issued, which we haven't mocked here.

**Comment by wicknicks** [comment]
@hachikuji responded to one of your comments, and updated the tests for others. Let me know what you think. Thanks!

File: core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala
Comment (wicknicks) [reply core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala]: added comment.

File: core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala
Comment (wicknicks) [reply core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala]: yes I was creating a fourth member. removed it now.

File: core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala
Comment (wicknicks) [reply core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala:962]: done.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java
Comment (hachikuji) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java]: One way to do it is to use 0 as the timeout in `joinGroupIfNeeded`. I tried this out locally and it seems to work.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java
Comment (hachikuji) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java]: nit: maybe a better name is `testPendingMemberShouldLeaveGroup`?

File: core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala
Comment (hachikuji) [inline core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala:934]: I think this test is good now. I verified that it fails without the fix in `onExpireHeartbeat`. The other case that is worth having a test for is if the pending member successfully rejoins. Similarly we should see the group transition to `CompletingRebalance`. The test cases will be mostly the same. The only difference is whether we advance the clock and expire or rejoin the pending member, so it might be possible to factor out the rest of the logic.

**Review by hachikuji — COMMENTED** [review]
Thanks @wicknicks . Just a couple more comments.

File: core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala
Comment (wicknicks) [reply core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala:934]: addressed this comment, @hachikuji. let me know what you think.

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala
Comment (hachikuji) [inline core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala]: Leaving the comment here for lack of a better location. We need to update `handleLeaveGroup` to handle pending members as well. It should follow logic similar to this. We remove the pending member, cancel the heartbeat, and check if the rebalance can continue. This would be a third case to check in addition to pending member expiration and pending member rejoin.

Comment (wicknicks) [reply core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala]: great catch. added the fix and a test.

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala
Comment (ijuma) [inline core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala:915]: Out of curiosity, why did we remove this?

Comment (hachikuji) [reply core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala:915]: Yeah, I think the message is useful. I'd even consider raising the verbosity to info. My only comment was to add groupId to it.

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala
Comment (hachikuji) [inline core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala]: The `is(Dead)` check should probably come first.

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala
Comment (ijuma) [reply core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala:915]: +1 to having this as `info`.

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala
Comment (hachikuji) [inline core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala]: Could we reuse `removeHeartbeatForLeavingMember`? We might need to add the `isLeaving` check to `tryCompleteHeartbeat`.

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala
Comment (hachikuji) [inline core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala]: We have two separate cases in which we now remove the pending member and check for rebalance completion. I think it might be worth having a `removePendingMemberAndUpdateGroup` for this logic to go along with `removeMemberAndUpdateGroup`.

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala
Comment (wicknicks) [reply core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala:915]: ok cool. I thought this was adding too much verbosity. I did find it useful when debugging (but this was my first time looking at these logs). I'll add a similar line in `handleLeaveGroup`.

**Review by hachikuji — COMMENTED** [review]
Just a few more comments. Thanks for all the effort over the weekend @wicknicks.

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala
Comment (wicknicks) [reply core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala:915]: message added back with info level and with groupID

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala
Comment (wicknicks) [reply core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala]: this has been added.

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala
Comment (wicknicks) [reply core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala]: we'd have to split the `group.has(memberId)` into a separate check then. Otherwise, for the pending member case, that check will trigger to true.

Comment (hachikuji) [reply core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala]: Yeah, that's fine. All we're doing is returning an error code, so the duplication seems harmless.

**Comment by wicknicks** [comment]
@hachikuji addressed your comments. thanks!

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: This docstring could be improved. Perhaps we can just mention that we will may have a memberId before we are part of a group generation.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java
Comment (hachikuji) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java:771]: Can we verify that the request is a LeaveGroupRequest?

**Review by hachikuji — APPROVED** [review]
Thanks, just a couple minor comments. LGTM overall.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java
Comment (wicknicks) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java:771]: done.

**Comment by wicknicks** [comment]
@hachikuji addressed your last two comments. thanks!

**Comment by hachikuji** [comment]
retest this please

**Comment by wicknicks** [comment]
@hachikuji looks like the tests passed. although the JDK 11 build finished running unit and integration test in 1h 39m?

**Comment by hachikuji** [comment]
Thanks for the patch. Merging to trunk and 2.2.

**Comment by abbccdda** [comment]
@wicknicks Sorry for late response Arjun. Thank you so much for fixing the leaving group logic!

**Pull request merged**
